lib/paper_trail/record_trail.rb

```diff
@@ -414,13 +410,29 @@ def add_transaction_id_to(data)
       data[:transaction_id] = PaperTrail.transaction_id
     end
 
+    # Given a HABTM association, returns an array of ids.
+    # @api private
+    def habtm_assoc_ids(habtm_assoc)
+      current = @record.send(habtm_assoc.name).to_a.map(&:id) # TODO: `pluck` would use less memory
```

Comment:
Upon closer inspection looks like you'll still need to invoke the association for removed and added records but the first line could be condensed using that helper